test/test_replication.py

```diff
@@ -0,0 +1,250 @@
+            "--start",
+        ],
+        stdout=subprocess.PIPE,
+    )
+    assert process.stdout is not None
+    bouncer.sql("INSERT INTO test_logical_rep_pg_recvlogical.table values (1)")
+    try:
+        assert process.stdout.readline().startswith(b"BEGIN ")
+        assert (
+            process.stdout.readline()
+            == b'table test_logical_rep_pg_recvlogical."table": INSERT: a[integer]:1\n'
+        )
+        assert process.stdout.readline().startswith(b"COMMIT ")
+    finally:
+        process.kill()
+        process.communicate(timeout=5)
+
+
+def test_physical_rep(bouncer):
+    connect_args = {
+        "dbname": "user_passthrough",
+        "replication": "yes",
+        "user": "postgres",
+        "application_name": "abc",
+        "options": "-c enable_seqscan=off",
+    }
+    # Starting in PG10 you can do SHOW commands
+    if PG_MAJOR_VERSION >= 10:
+        with pytest.raises(
+            psycopg.errors.FeatureNotSupported,
+            match="cannot execute SQL commands in WAL sender for physical replication",
+        ):
+            bouncer.test(**connect_args)
+        assert bouncer.sql_value("SHOW application_name", **connect_args) == "abc"
+        assert bouncer.sql_value("SHOW enable_seqscan", **connect_args) == "off"
+    bouncer.sql("IDENTIFY_SYSTEM", **connect_args)
+    # Do a normal connection to the same pool, to ensure that that doesn't
+    # break anything
+    bouncer.test(dbname="user_passthrough", user="postgres")
+    bouncer.sql("IDENTIFY_SYSTEM", **connect_args)
+
+
+def test_physcal_rep_unprivileged(bouncer):
+    with bouncer.log_contains(
+        r"no pg_hba.conf entry for replication connection from host"
+    ), bouncer.log_contains(
+        r"closing because: login failed \(age", times=2
+    ), pytest.raises(
+        psycopg.OperationalError, match=r"login failed"
+    ):
+        bouncer.test(replication="yes")
+
+
+@pytest.mark.skipif("PG_MAJOR_VERSION < 10", reason="pg_receivewal was added in PG10")
+def test_physical_rep_pg_receivewal(bouncer, tmp_path):
+    bouncer.default_db = "user_passthrough"
+    bouncer.create_physical_replication_slot("test_physical_rep_pg_receivewal")
+    wal_dump_dir = tmp_path / "wal-dump"
+    wal_dump_dir.mkdir()
+
+    process = subprocess.Popen(
+        [
+            "pg_receivewal",
+            "--dbname",
+            bouncer.make_conninfo(),
+            "--slot=test_physical_rep_pg_receivewal",
+            "--directory",
+            str(wal_dump_dir),
+        ],
+    )
+    time.sleep(3)
+
+    if WINDOWS:
+        process.terminate()
+    else:
+        process.send_signal(signal.SIGINT)
+    process.communicate(timeout=5)
+
+    if WINDOWS:
+        assert process.returncode == 1
+    else:
+        assert process.returncode == 0
+
+    children = list(wal_dump_dir.iterdir())
+    assert len(children) > 0
+
+
+@pytest.mark.skipif(
+    "PG_MAJOR_VERSION < 10", reason="pg_basebackup on PG9.6 fails with a weird error"
```

Comment:
I'm curious about the "weird error". Is it a random failure? Or it is because of a behavior change (it streams WAL by default)?

Reply:
I took a bit closer look and I was able to fix it by the flag `-X stream`.